File: Java/Behavioral/StatePattern.java
package Behavioral;

interface State {
    void doAction(Context context);
}

class StartState implements State {
    public void doAction(Context context) {
        System.out.println("Player is in Start State");
        context.setState(this);
    }
}

class StopState implements State {
    public void doAction(Context context) {
        System.out.println("Player is in Stop State");
        context.setState(this);
    }
}

class Context {
    private State state;

    public void setState(State state) {
        this.state = state;
    }

    public State getState() {
        return state;
    }
}

public class StatePattern {
    public static void main(String[] args) {
        Context context = new Context();

        StartState startState = new StartState();
        startState.doAction(context);

        System.out.println(context.getState().getClass().getName());

        StopState stopState = new StopState();
        stopState.doAction(context);

        System.out.println(context.getState().getClass().getName());
    }
}

File: Python/Behavioral/StatePattern.py
class State:
    def do_action(self, context):
        pass

class StartState(State):
    def do_action(self, context):
        print("Player is in Start State")
        context.state = self

class StopState(State):
    def do_action(self, context):
        print("Player is in Stop State")
        context.state = self

class Context:
    def __init__(self):
        self.state = None

    def set_state(self, state):
        self.state = state

    def get_state(self):
        return self.state

# Usage
context = Context()

start_state = StartState()
start_state.do_action(context)

print(f"Current State: {context.get_state().__class__.__name__}")

stop_state = StopState()
stop_state.do_action(context)

print(f"Current State: {context.get_state().__class__.__name__}")
